tests/test-dladdr.c

```diff
@@ -0,0 +1,163 @@
+        int addr_match  = addr == info.dli_saddr;
+        passed = (expected_result == Pass && sym_match && addr_match)
+                 || (expected_result == PassWithoutSymbol && addr_match)
+                 || (expected_result == Fail && !(sym_match && addr_match));
+        printf( "check address %p with has symbol '%s' -> %s\n",addr, addrsym, passed ? "passed" : "failed" );
+        if (verbose)
+            print_dl_info( &info );
+    }
+    return !result;
+}
+
+/**
+ * @brief return address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @return pointer to symbol address or NULL
+ */
+void *get_symbol_address( char *libname, char *sym )
+{
+    void *library = NULL;
+    void *addr = NULL;
+
+    library = dlopen( libname, RTLD_GLOBAL );
+    if ( library == NULL )
+    {
+        fprintf( stderr, "could not open '%s'\n", libname );
+        return NULL;
+    }
+
+    addr = dlsym( library, sym );
+    dlclose( library );
+    if (!addr)
+        fprintf( stderr, "could not get address for symbol '%s'\n", sym );
+    return addr;
+}
+
+/**
+ * @brief check address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @param should_match result should match the given values
+ * @return 0 check passed
+ * @return 1 check failed
+ */
+int check_dladdr_by_dlopen( char *libname, char *sym, int should_match )
+{
+    void *addr = get_symbol_address( libname, sym);
+    if (!addr)
+        return 1;
+    return check_dladdr( addr, sym, should_match );
+}
+
+#ifdef _WIN32
+#define HMODULE void*
+#define HANDLE void*
+#define LPCVOID void*
+#define LPCSTR char*
+#define DWORD long
+#define SIZE_T long
+
+#ifdef _MSC_VER
+#define STDCALL __stdcall
+#else
+#define STDCALL __attribute__((__stdcall__))
+#endif
+
+/* link to import thunk */
+HMODULE STDCALL GetModuleHandleA (LPCSTR lpModuleName);
+/* link to directly to iat */
+__declspec(dllimport) HMODULE STDCALL LoadLibraryExA (LPCSTR lpLibFileName, HANDLE hFile, DWORD dwFlags);
+SIZE_T STDCALL VirtualQuery (LPCVOID lpAddress, LPCVOID lpBuffer, SIZE_T dwLength);
+#define PassWithoutSymbolOnWin PassWithoutSymbol
+#else
+#define PassWithoutSymbolOnWin Pass
+#endif
+
+
+int main(int argc, char **argv)
+{
+    int  result = 0;
+    if (argc == 2)
+        verbose = 1;
+
+    /* null pointer */
+    result = check_dladdr((void*)0, "null" , NoInfo);
+    /* invalid pointer */
+    result |= check_dladdr((void*)0x125, "invalid pointer" , Fail);
+    /* -ldl */
+    result |= check_dladdr( ((void*)dladdr), "dladdr" , Pass );
+    /* -ldl */
+    result |= check_dladdr( (void*)dlopen, "dlopen", Pass );
+    /* -lglibc */
+    result |= check_dladdr( (void*)vsnprintf, "vsnprintf", PassWithoutSymbolOnWin );
+    /* test-dladdr */
+    result |= check_dladdr( (void*)main, "main", PassWithoutSymbolOnWin );
```

Comment:
I compiled this test with mingw gcc `-Wl,--export-all-symbols` parameter and then in resulted EXE file is `main` symbol present, see objdump:

```
$ objdump --private-headers test-dladdr.exe
...
There is an export table in .edata at 0x406000

The Export Tables (interpreted .edata section contents)

Export Flags                    0
Time/Date stamp                 5f8c34d6
Major/Minor                     0/0
Name                            0000605a test-dladdr.exe
Ordinal Base                    1
Number in:
        Export Address Table            00000005
        [Name Pointer/Ordinal] Table    00000005
Table Addresses
        Export Address Table            00006028
        Name Pointer Table              0000603c
        Ordinal Table                   00006050

Export Address Table -- Ordinal Base 1
        [   0] +base[   1] 1550 Export RVA
        [   1] +base[   2] 17a0 Export RVA
        [   2] +base[   3] 16f0 Export RVA
        [   3] +base[   4] 2910 Export RVA
        [   4] +base[   5] 1510 Export RVA

[Ordinal/Name Pointer] Table
        [   0] check_dladdr
        [   1] check_dladdr_by_dlopen
        [   2] get_symbol_address
        [   3] main
        [   4] print_dl_info
...
```

But dladdr does not see this symbol:
```
check address 00402910 with has symbol 'main' -> passed
filename: ...\test-dladdr.exe base: 00400000 symbol name: '(null)' addr: 00402910
```

So it seems that `dladdr` implementation cannot resolve symbols from main executable in case they are exported. What do you think, is there bug in `dladdr` implementation? Or it is expected and parsing is not implemented yet?

Reply:
> So it seems that dladdr implementation cannot resolve symbols from main executable in case they are exported

dladdr only checks the import table, not the exported table. It may be possible to check the export table too, but I haven't looked yet.